ul/src/pdu/test.rs

```diff
@@ -0,0 +1,143 @@
+use crate::pdu::reader::*;
+use crate::pdu::writer::*;
+use crate::pdu::*;
+use byteordered::byteorder::WriteBytesExt;
+use std::io::Cursor;
+
+#[test]
+fn can_write_chunks_with_preceding_u32_length() -> crate::error::Result<()> {
+    let mut bytes = vec![0u8; 0];
+    crate::pdu::writer::write_chunk_u32(&mut bytes, |writer| {
+        writer.write_u8(0x02)?;
+        crate::pdu::writer::write_chunk_u32(writer, |writer| {
+            writer.write_u8(0x03)?;
+            Ok(())
+        })?;
+        Ok(())
+    })?;
+
+    assert_eq!(bytes.len(), 10);
+    assert_eq!(bytes, &[0, 0, 0, 6, 2, 0, 0, 0, 1, 3]);
+
+    Ok(())
+}
+
+#[test]
+fn can_write_chunks_with_preceding_u16_length() -> crate::error::Result<()> {
+    let mut bytes = vec![0u8; 0];
+    crate::pdu::writer::write_chunk_u16(&mut bytes, |writer| {
+        writer.write_u8(0x02)?;
+        crate::pdu::writer::write_chunk_u16(writer, |writer| {
+            writer.write_u8(0x03)?;
+            Ok(())
+        })?;
+        Ok(())
+    })?;
+
+    assert_eq!(bytes.len(), 6);
+    assert_eq!(bytes, &[0, 4, 2, 0, 1, 3]);
+
+    Ok(())
+}
+
+#[test]
+fn can_read_write_associate_rq() -> crate::error::Result<()> {
+    let association_rq = PDU::AssociationRQ {
+        protocol_version: 2,
+        calling_ae_title: "calling ae".to_string(),
+        called_ae_title: "called ae".to_string(),
+        application_context_name: "application context name".to_string(),
+        presentation_contexts: vec![
+            PresentationContextProposed {
+                id: 1,
+                abstract_syntax: "abstract 1".to_string(),
+                transfer_syntaxes: vec!["transfer 1".to_string(), "transfer 2".to_string()],
+            },
+            PresentationContextProposed {
+                id: 3,
+                abstract_syntax: "abstract 2".to_string(),
+                transfer_syntaxes: vec!["transfer 3".to_string(), "transfer 4".to_string()],
+            },
+        ],
+        user_variables: vec![
+            UserVariableItem::ImplementationClassUID("class uid".to_string()),
+            UserVariableItem::ImplementationVersionName("version name".to_string()),
+            UserVariableItem::MaxLength(23),
+        ],
+    };
+
+    let mut bytes = vec![0u8; 0];
+    write_pdu(&mut bytes, &association_rq)?;
+
+    let result = read_pdu(&mut Cursor::new(&bytes), DEFAULT_MAX_PDU)?;
+
+    if let PDU::AssociationRQ {
+        protocol_version,
+        calling_ae_title,
+        called_ae_title,
+        application_context_name,
+        presentation_contexts,
+        user_variables,
+    } = result
+    {
+        assert_eq!(protocol_version, 2);
+        assert_eq!(calling_ae_title, "calling ae".to_string());
```

Comment:
I suppose that `String` values can be compared with string slices directly (if the compiler complains, prepend `calling_ae_title` with a `&`).

```suggestion
        assert_eq!(calling_ae_title, "calling ae");
```